### PR TITLE
Try to fix some flaky tests

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
We have been suffering from flaky tests for a while. Let's fix some of them.

Modification:

* `SendFromOwned` has changed its state tracking to use and `int` instead of a `boolean`, since we hid a suspected mis-compile with Java 17 Graal, where the result of the `getAndSet` would appear to return the update value rather than the existing value.
* Removed use of timeouts in `WebSocketClientHandshakerTest` because it's using `EmbeddedChannel`, so there's no concurrency involved, and the test should not be able to stall.
* Change `InboundHttp2ToHttpAdapterTest` to not use mocking, but instead now uses special-purpose and thread-safe `Listeners` for communicating events from within the pipeline out to the test for verification. Mocks are not expected to be thread-safe, so this fixes a correctness issue and also makes the test run faster.
* Fix a problem with how `SocketTestPermutation` generated unique domain socket addresses; unique temporary files were created, and then deleted, to be created again as domain sockets. This created a window for the names to race and multiple tests could end up accidentally sharing domain sockets. This has been changed so that we now create a unique _directory_ per JVM process, and then use a shared atomic counter for naming files within that directory. This guarantees that no two tests will end up sharing a domain socket.
* The `SocketConnectionAttemptTest` had a race between an `await` in the test thread, and the starting of the event-loop thread and the following timeout of a connect attempt. The test has been changed to now just `sync` on the future, so there's no race. The timeout has also been reduced, because the connect attempt will eventually fail regardless, and there's no reason to wait for 2 seconds for this to occur.
* The `TrafficShapingHandlerTest` has been made to run isolated, and mandate SAME_THREAD execution mode, because it relies on `System.nanoTime` for measuring throughput or something like that, and then asserts on that value. This makes the test sensitive to the performance of the system, so it would often fail if other tests are running concurrently with it.
* The native transport tests now all use identical `junit-platform.properties` files.
* A reentrancy bug has been fixed in how `DefaultChannelHandlerContext` updates the pipeline pending outbound bytes. The pending bytes of the handler would be saved before a handler action, and queried again after, and the difference would be used to update the pipeline pending bytes. This had problems with reentrancy, which we tried to cope with using the `-1` special value, but it was not enough; on nested handler calls, the first call to `updatePendingBytesIfNeeded` would pair with the first call to `saveCurrentPendingBytesIfNeeded`, which would leave other such calls (when reentrant) not pairing in a correct way, and we could loose updates or do double-counting. This has been fixes by making the implementation reentrancy oblivious; we now just track the last observed pending bytes of the handler, and update whenever there's a difference. Basically, saving and updating are no longer two separate steps, but instead always happen together as a single operation. When there is only a single operation, we no longer have pairs that need to match, and reentrancy is no longer an issue.
* The `LocalChannel.doWriteNow` no longer specializes for `ResourceSupport` objects to call `acquire` on them. The `acquire` method is not thread safe. This lead to issues where we incremented the reference count in the sending thread, the receiving thread then obtained the object before the reference count was brought back down in the sending thread, and we could observe that e.g. `Buffer` objects were "not owned" when we tried to access them in the receiving thread. The fix is to just not do that, and instead leave the transfer of ownership to `Resource.send()` and `Send.receive()`, which is what those two were designed for.
* The `LocalChannelTest` no longer reuses `LocalAddress` objects, which would be a problem if the tests ran concurrently. Now every test get their own unique address to play with, and we no longer risk test isolation failures.

Result:
Our build should be more stable now.

### CODE REVIEW

**Comment by chrisvest** [comment]
This doesn't solve the real problem; just makes it less likely to bite any given PR. We still see these tests fail for the NIO transport, which isn't running anything in parallel.

File: testsuite/src/main/java/io/netty5/testsuite/transport/socket/SocketTestPermutation.java [hunk 1]
```diff
@@ -203,18 +204,39 @@ public List<BootstrapFactory<Bootstrap>> datagramSocket() {
...
+        return new DomainSocketAddress(DomainSocketDirectory.nextSocketName());
+    }
+
+    /**
+     * Separate class so we only initialize this field if we really need it.
+     */
+    private static final class DomainSocketDirectory {
+        public static final int MAX_PATH_LEN = 128;
+        public static final AtomicInteger COUNTER = new AtomicInteger();
+        public static final File DIR;
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty5/testsuite/transport/socket/SocketTestPermutation.java:218]:
```suggestion
        private static final int MAX_PATH_LEN = 128;
        private static final AtomicInteger COUNTER = new AtomicInteger();
        private static final File DIR;
```

File: testsuite/src/main/java/io/netty5/testsuite/transport/socket/SocketTestPermutation.java [hunk 2]
```diff
@@ -203,18 +204,39 @@ public List<BootstrapFactory<Bootstrap>> datagramSocket() {
...
+        static {
+            try {
+                int maxLen = MAX_PATH_LEN - File.separator.length() - 2 * Integer.BYTES;
+                File file = null;
+                do {
+                    if (file != null && !file.delete()) {
+                        throw new IOException("failed to delete: " + file);
+                    }
+                    file = PlatformDependent.createTempDirectory("NETTY-UDS", null).getAbsoluteFile();
+                } while (file.toString().length() > maxLen);
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty5/testsuite/transport/socket/SocketTestPermutation.java:229]: Let's ensure we not loop forever
```suggestion
                int maxTries = 64;
                do {
                    if (maxTries-- == 0) {
                        throw new ExceptionInInitializerError("Unable to create temporary file with path length <= " + maxLen);
                    }
                    if (file != null && !file.delete()) {
                        throw new IOException("failed to delete: " + file);
                    }
                    file = PlatformDependent.createTempDirectory("NETTY-UDS", null).getAbsoluteFile();
                } while (file.toString().length() > maxLen);
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

**Comment by chrisvest** [comment]
@gradinac The Java 17 Graal build found something interesting that looks very much like a Graal JIT compiler bug around VarHandles and x86_64 atomics. I'm guessing you know who to tell about this.

The stack trace of interest looks like this:

```
Caused by: java.lang.IllegalStateException: This object has already been received.
	at io.netty5.buffer.internal.SendFromOwned.gateReception(SendFromOwned.java:54)
	at io.netty5.buffer.internal.SendFromOwned.receive(SendFromOwned.java:46)
	at io.netty5.buffer.BufferInputStream.<init>(BufferInputStream.java:61)
	at io.netty5.handler.ssl.SslContext.getCertificatesFromBuffers(SslContext.java:1247)
	at io.netty5.handler.ssl.SslContext.toX509Certificates(SslContext.java:1232)
	at io.netty5.handler.ssl.SslContextBuilder.keyManager(SslContextBuilder.java:378)
	... 110 more
```

The exception comes when `gateReception` fails to flip a volatile boolean field from `false` to `true`, via `getAndSet` through a `VarHandle`: https://github.com/netty/netty/blob/45ead82fba89238a703ce1afca1730dc64a246f1/buffer/src/main/java/io/netty5/buffer/internal/SendFromOwned.java#L52-L61

The field is local to the `SendFromOwned` object, and keeps its initial `false` value until the object is received or closed: https://github.com/netty/netty/blob/45ead82fba89238a703ce1afca1730dc64a246f1/buffer/src/main/java/io/netty5/buffer/internal/SendFromOwned.java#L34

So the happy-path is that the `getAndSet` will return `false`, and atomically set the field value to `true`, preventing any further receives in a thread-safe way. A `Send` can only be received once.

The thing is, the `Buffer`, the `SendFromOwned`, the `send()` and `receive()` calls all happen in the same thread, and all objects are local to that thread.

The buffers are allocated in the `PemReader.readCertificates()` call: https://github.com/netty/netty/blob/main/handler/src/main/java/io/netty5/handler/ssl/SslContext.java#L1228-L1233

Then sent to the `BufferInputStream` constructor purely as a matter of buffer ownership transfer: https://github.com/netty/netty/blob/main/handler/src/main/java/io/netty5/handler/ssl/SslContext.java#L1242-L1255

And received in the constructor, in the same thread: https://github.com/netty/netty/blob/45ead82fba89238a703ce1afca1730dc64a246f1/buffer/src/main/java/io/netty5/buffer/BufferInputStream.java#L60-L66

The `SendFromOwned` object was created by the `send()` call in the `getCertificatesFromBuffers()` call above. The `send()` is implemented like this: https://github.com/netty/netty/blob/45ead82fba89238a703ce1afca1730dc64a246f1/buffer/src/main/java/io/netty5/buffer/internal/ResourceSupport.java#L136-L150

As you can see, each `send()` call creates a new `SendFromOwned` object. The buffer is also put in a state that prevents more `Send` objects from being created, though that's not relevant to this issue.

At no point do any of these objects get shared with other threads, and each buffer is only sent once, and each `Send` is immediately received on the same thread. So there's no way the `received` could be anything other than `false`, yet the `getAndSet` call returns `true` and the exception is thrown.

**Review by normanmaurer — APPROVED** [review]

**Comment by gradinac** [comment]
@chrisvest Thank you very much for looking into this! :) I've also looked into the gate logs to verify, but just to double check, this happens in the Graal JIT mode, not on native-image?

Could you, if you have the time, open an issue under [oracle/graal](https://github.com/oracle/graal): https://github.com/oracle/graal/issues/new?assignees=&labels=bug&template=5_issues_other.md&title=
One thing that would really help to debug it would be a standalone reproducer. It would be perfect if we had one, but if that's not possible, the set of commands that would run just the above test should also help. I assume it would be something like:

To setup:
```bash
export JAVA_HOME=<path-to-graalvm>
git clone https://github.com/netty/netty.git
cd netty
git fetch origin main
git checkout main
mvn -DskipTests=true install
cd handler
```
To run just this test:
```bash
mvn -Dtest=ParameterizedSslHandlerTest.testCompositeBufSizeEstimationGuaranteesSynchronousWrite test
```

**Comment by chrisvest** [comment]
@gradinac 
> Could you, if you have the time, open an issue under

Employer says I'm not allowed, so no 🤷 

Reproducer instructions looks correct, except the command to run the test needs to look like this:

```
mvn -Dtest=ParameterizedSslHandlerTest#testCompositeBufSizeEstimationGuaranteesSynchronousWrite test -pl handler
```

**Pull request merged**